components/script/script_thread.rs

```diff
@@ -1367,107 +1372,96 @@ impl ScriptThread {
-        // regular rendering opportunities from the compositor (or fake animation frame
-        // ticks). In this case, don't schedule an opportunity, just wait for the next
-        // one.
-        if self.documents.borrow().iter().any(|(_, document)| {
+    /// Schedule a rendering update ("update the rendering"), if necessary. This
+    /// can be necessary for a couple reasons. For instance, when the DOM
+    /// changes a scheduled rendering update becomes necessary if one isn't
+    /// scheduled already. Another example is if rAFs are running but no display
+    /// lists are being produced. In that case the [`ScriptThread`] is
+    /// responsible for scheduling animation ticks.
+    fn maybe_schedule_rendering_opportunity_after_ipc_message(
+        &self,
+        built_any_display_lists: bool,
+    ) {
+        let needs_rendering_update = self
+            .documents
+            .borrow()
+            .iter()
+            .any(|(_, document)| document.needs_rendering_update());
+        let running_animations = self.documents.borrow().iter().any(|(_, document)| {
             document.is_fully_active() &&
+                !document.window().throttled() &&
                 (document.animations().running_animation_count() != 0 ||
                     document.has_active_request_animation_frame_callbacks())
-        }) {
+        });
+
+        // If we are not running animations and no rendering update is
+        // necessary, just exit early and schedule the next rendering update
+        // when it becomes necessary.
+        if !needs_rendering_update && !running_animations {
             return;
         }
 
-        let Some((_, document)) = self.documents.borrow().iter().find(|(_, document)| {
-            document.is_fully_active() &&
-                !document.window().layout_blocked() &&
-                !document.restyle_reason().is_empty()
-        }) else {
+        // If animations are running and a reflow in this event loop iteration
+        // produced a display list, rely on the renderer to inform us of the
+        // next animation tick / rendering opportunity.
+        if running_animations && built_any_display_lists {
             return;
+        }
+
+        // There are two possibilities: rendering needs to be updated or we are
+        // scheduling a new animation tick because animations are running, but
+        // not changing the DOM. In the later case we can wait a bit longer
+        // until the next "update the rendering" call as it's more efficient to
+        // slow down rAFs that don't change the DOM.
+        let animation_delay = if running_animations && !needs_rendering_update {
+            Duration::from_millis(30)
+        } else {
+            // 20 milliseconds is used here in order to allow any renderer-based
+            // animation ticks to arrive first.
+            //
+            // TODO: Should this be faster to reduce update latency?
+            Duration::from_millis(20)
```

Comment:
Can we write FPS in comments, as I think this is more relevant in this context than ms.